### PR TITLE
Fix tag filtering in search entry

### PR DESCRIPTION
**Pull request by bimlas**
Some issues introduced in #954 (https://github.com/BoostIO/Boostnote/pull/954#issuecomment-336344915):

- search by `#tag1 #tag2` returns the results of only `#tag2`
- search by `#tag1 content` is as same as AND search by `#tag1 content` (and `#tag1` is regarded a word, not a tag)
- search by `content #tag1` returns the results of only `#tag1`

This commit fixing these:

- search by `#tag1 #tag2` returns the results of only `#tag2`

  screencast

- search by `#tag1 content` is as same as AND search by `#tag1 content` (and `#tag1` is regarded a word, not a tag)

  screencast

- search by `content #tag1` returns the results of only `#tag1`

  screencast

NOTE: the examples works without `#` character too, because `findByWord()` checks the tags too.

### Related issues:
* #1499 
* #1683

### CODE REVIEW

**Comment by bimlas** [comment]
Travis CI says:

> SyntaxError: Unexpected token } in JSON at position 108
https://travis-ci.org/BoostIO/Boostnote/builds/355745053#L513

If I'm right, the code syntax is correct and the [related test](https://github.com/bimlas/Boostnote/blob/5f5a7880a661df9accb5a4483c18724b38535f0b/tests/lib/search-test.js#L33) works when tested manually. What's the problem? And why is this [related to *.boostnoterc*](https://travis-ci.org/BoostIO/Boostnote/builds/355745053#L530)?

**Comment by Rokt33r** [comment]
You can ignore the log. Our tests are a little noisy. 😉

**Comment by bimlas** [comment]
Ok, but why the tests are failing if everything seems to be OK? The branch cannot be merged until all of the tests passing.

**Comment by Rokt33r** [comment]
@bimlas Check the test cases. You've changed the specification. So, there should be some conflicts.

**Comment by bimlas** [comment]
I see that test fails but when I trying out manually, it works. I overlooked the diff and it has to work... :confused: 

Before:

image

After:

image

As I see the [warning about syntax error in JSON and *.boostnoterc* in the log](https://travis-ci.org/BoostIO/Boostnote/builds/355729522#L503-L520) exists on `master` too, so it's not related to this PR.

Anyway: I'm trying to fix it. :+1:

**Comment by bimlas** [comment]
Got it - forced push.

**Comment by bimlas** [comment]
I'm not sure, but it may slow down the search - somebody with a big repo (with tags) should check it.

**Comment by bimlas** [comment]
Please don't merge it for a while: I think the search methoud could be faster.

```javascript
    if (block.match(/^#.+/)) {
      foundNotes = findByTag(foundNotes.slice(0), block).concat(findByWord(foundNotes.slice(0), block))
    } else {
      foundNotes = findByWord(foundNotes, block)
    }
```
Searching for a `#tag` parsing the tags in the note's tag list **and** the content. But the "simple" word searching doing the same:

```javascript
function findByWord (notes, block) {
  const regExp = new RegExp(_.escapeRegExp(block), 'i')
  return notes.filter((note) => {
    if (_.isArray(note.tags) && note.tags.some((_tag) => {
      return _tag.match(regExp)
    })) {
      return true
    }
    if (note.type === 'SNIPPET_NOTE') {
      return note.description.match(regExp)
    } else if (note.type === 'MARKDOWN_NOTE') {
      return note.content.match(regExp)
    }
    return false
  })
}
```

So this is a code duplication, thus it can be refactored to one method which could be called only once -> the searching would speed up. Please give me some time to do it.

**Comment by bimlas** [comment]
I'm done - I don't experience speed improvement, but the code became simpler. :D

**Comment by bimlas** [comment]
Just added another test: searching has to find tags without hash symbol too (`tag` instead of
`#tag`).

**Comment by bimlas** [comment]
Ping @Rokt33r I'm done, the branch may be merged.

File: browser/lib/search.js [hunk 1]
```diff
@@ -4,39 +4,30 @@ export default function searchFromNotes (notes, search) {
...
-    return note.tags.some((_tag) => {
-      return _tag.match(regExp)
-    })
-  })
-}
-
-function findByWord (notes, block) {
-  const regExp = new RegExp(_.escapeRegExp(block), 'i')
+function findByWordOrTag (notes, block) {
+  let tag = block
```
[the remark this answers is not in the document]
Comment (bimlas) [reply browser/lib/search.js]: Because this is the behaviour (in the current release too): the searching looking for the word in tags too.

**Comment by bimlas** [comment]
@Rokt33r Don't you like that search looks for word in tags too? I think it has to be the behaviour: tags can be used to find content precisely (`#javascript #string #print` for example), while "text search" is used every content related to a given word (even if the note is tagged with this word).

**Comment by Rokt33r** [comment]
I'd just needed more time to think. I assume you're right.

The reason why I should postpone merging this is that it fixes more than one thing. 😵

**Comment by bimlas** [comment]
> it fixes more than one thing

Do you mean the commits or the PR?

**Comment by bimlas** [comment]
How should I split it? It does not adds behaviour, but fixes it + prevents the breaking of it (by adding test cases).

**Comment by Rokt33r** [comment]
I think you could submit each fix.

**Pull request merged**